# Задача предсказания. Линейная регрессия

Предсказание числового признака называется **задачей регрессии**

Предсказываемый признак называется **целевым**



In [16]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [17]:
import pandas as pd
import seaborn as sns
from sklearn.linear_model import LinearRegression
# подключить библиотеку с линейной регрессией

In [18]:
df=pd.read_csv('/content/drive/MyDrive/Colab Notebooks/students.csv', delimiter=',')

In [19]:
df_cut=df[['Growth','Shoe size']]
df_cut=df_cut.dropna()

In [ ]:
sns.scatterplot(data=df_cut, x='Shoe size', y='Growth')

Гипотеза: рост линейно зависит от размера обуви.
Growth=a*ShSize+b
коэффициента a,b неизвестны
Модель линейной регрессии умеет находить оптимальные коэффициенты

In [21]:
lienearRegression = LinearRegression()
results=lienearRegression.fit(df_cut['Shoe size'].values.reshape(-1, 1), y=df_cut['Growth'].values)

In [ ]:
df_cut['Shoe size'].values.reshape(-1,1)

In [23]:
results.coef_, results.intercept_

(array([2.77717058]), 60.36581125736733)

growth=a*Shsize+b

growth=2.7*Shsize+60


In [24]:
df_cut['Predicted Growth']=results.predict(df_cut['Shoe size'].values.reshape(-1, 1))

In [25]:
from sklearn.metrics import mean_absolute_error
mean_absolute_error(df_cut['Growth'], df_cut['Predicted Growth'])

3.450698681998386

In [ ]:
df_cut.head(20)

In [27]:
df_cut=df[['MIddle and index finger','Middle and ring finger','Middle and little finger']]

In [28]:
results=lienearRegression.fit(df_cut[['Middle and ring finger','Middle and little finger']].values.reshape(-1, 2), y=df_cut['MIddle and index finger'].values)
results.intercept_,results.coef_

(7.436145518798954, array([0.10731496, 0.04389948]))

MIddle and index finger=0.107 Middle and ring finger+0.04Middle and little finger+7.44


In [29]:
df_cut['Predicted']=results.predict(df_cut[['Middle and ring finger','Middle and little finger']].values.reshape(-1, 2))

<ipython-input-29-fcb0a82fc094>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cut['Predicted']=results.predict(df_cut[['Middle and ring finger','Middle and little finger']].values.reshape(-1, 2))


In [30]:
mean_absolute_error(df_cut['Predicted'], df_cut['MIddle and index finger'])

3.094567105647603

# Проверка точности модели на новой выборке

До этого момента мы тренировали модель и проверяли ее качество (mean_absolute_error) на одних и тех же данных.

Вообще говоря, это читерство, так как модель была построена **с участием данных**, на которых она потом проверяется.

Поэтому в анализе данных поступают так:

У вас есть **две** таблицы: **тренировочная и тестовая выборка.**

Модель строится по **тренировочной** выборке. Качество ее проверяется по **тестовой** выборке.

Сейчас мы это реализуем. Загружаем второую таблицу со студентами, которая не использовалась ранее:

In [31]:
df_test=pd.read_csv('/content/drive/MyDrive/Colab Notebooks/students_test.csv', delimiter=',')
df_test_cut=df_test[['MIddle and index finger','Middle and ring finger','Middle and little finger']]

Внимание: через функцию predict прогоняем объекты **тестовой** выборки

In [32]:
df_test_cut['Predicted']=results.predict(df_test_cut[['Middle and ring finger','Middle and little finger']].values.reshape(-1, 2))

<ipython-input-32-84dcb0dfd551>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_test_cut['Predicted']=results.predict(df_test_cut[['Middle and ring finger','Middle and little finger']].values.reshape(-1, 2))


In [ ]:
df_test_cut.head()

In [ ]:
from sklearn.metrics import mean_absolute_error
mean_absolute_error(df_test['Growth'], df_test_cut['Predicted'])

In [56]:
df_cut=df[['Age','Year of birth', 'Month of birthday']]
lienearRegression = LinearRegression()
results=lienearRegression.fit(df_cut[['Year of birth', 'Month of birthday']].values.reshape(-1, 2), y=df_cut['Age'].values)

In [57]:
df_test_cut=df_test[['Age','Year of birth', 'Month of birthday']]
df_test_cut['Predicted Age']=results.predict(df_test_cut[['Year of birth', 'Month of birthday']].values.reshape(-1, 2))
df_test_cut

<ipython-input-57-3b3e664e7118>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_test_cut['Predicted Age']=results.predict(df_test_cut[['Year of birth', 'Month of birthday']].values.reshape(-1, 2))


Age  Year of birth  Month of birthday  Predicted Age
0    22           2000                  6      20.621444
1    20           2002                  3      20.318359
2    21           2001                  6      20.619989
3    21           2001                  7      20.720048
4    22           2000                  5      20.521385
..  ...            ...                ...            ...
93   20           2002                  2      20.218300
94   22           2000                  4      20.421326
95   22           2000                  4      20.421326
96   20           2001                  9      20.920165
97   20           2002                  3      20.318359

[98 rows x 4 columns]

In [55]:
from sklearn.metrics import mean_absolute_error
mean_absolute_error(df_test['Age'], df_test_cut['Predicted Age'])

1.6054960000660923